File: cap05/src/main/java/app/car/cap05/interfaces/incoming/mapping/TravelRequestMapper.java
package app.car.cap05.interfaces.incoming.mapping;

import app.car.cap05.domain.Passenger;
import app.car.cap05.domain.PassengerRepository;
import app.car.cap05.domain.TravelRequest;
import app.car.cap05.interfaces.incoming.PassengerAPI;
import app.car.cap05.interfaces.incoming.input.TravelRequestInput;
import app.car.cap05.interfaces.incoming.output.TravelRequestOutput;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.hateoas.EntityModel;
import org.springframework.hateoas.Link;
import org.springframework.hateoas.server.mvc.WebMvcLinkBuilder;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Component;
import org.springframework.web.server.ResponseStatusException;

import java.util.List;

@Component
public class TravelRequestMapper {

    @Autowired
    private PassengerRepository passengerRepository;

    public TravelRequest map(TravelRequestInput input) {

        Passenger passenger = passengerRepository.findById(input.getPassengerId())
                .orElseThrow(() -> new ResponseStatusException(HttpStatus.NOT_FOUND));

        TravelRequest travelRequest = new TravelRequest();
        travelRequest.setOrigin(input.getOrigin());
        travelRequest.setDestination(input.getDestination());
        travelRequest.setPassenger(passenger);

        return travelRequest;

    }

    public TravelRequestOutput map(TravelRequest travelRequest) {
        TravelRequestOutput travelRequestOutput = new TravelRequestOutput();

        travelRequestOutput.setCreationDate(travelRequest.getCreationDate());
        travelRequestOutput.setDestination(travelRequest.getDestination());
        travelRequestOutput.setId(travelRequest.getId());
        travelRequestOutput.setOrigin(travelRequest.getOrigin());
        travelRequestOutput.setStatus(travelRequest.getStatus());

        return travelRequestOutput;
    }

    public EntityModel<TravelRequestOutput> buildOutputModel(TravelRequest travelRequest, TravelRequestOutput output) {
        EntityModel<TravelRequestOutput> model = EntityModel.of(output);

        Link passengerLink = WebMvcLinkBuilder
                .linkTo(PassengerAPI.class)
                .slash(travelRequest.getPassenger().getId())
                .withRel("passenger")
                .withTitle(travelRequest.getPassenger().getName());
        model.add(passengerLink);
        return model;
    }

    public List<EntityModel<TravelRequestOutput>> buildOutputModel(List<TravelRequest> requests) {
        return requests.stream().map(tr -> buildOutputModel(tr, map(tr))).toList();
    }

}

File: cap09/src/main/kotlin/app/car/car09/interfaces/incoming/mapping/TravelRequestMapper.kt
package app.car.car09.interfaces.incoming.mapping

import app.car.car09.domain.Passenger
import app.car.car09.domain.PassengerRepository
import app.car.car09.domain.TravelRequest
import app.car.car09.interfaces.incoming.PassengerAPI
import app.car.car09.interfaces.incoming.input.TravelRequestInput
import app.car.car09.interfaces.incoming.output.TravelRequestOutput
import org.springframework.hateoas.EntityModel
import org.springframework.hateoas.Link
import org.springframework.hateoas.server.mvc.WebMvcLinkBuilder
import org.springframework.http.HttpStatus
import org.springframework.stereotype.Component
import org.springframework.web.server.ResponseStatusException

@Component
class TravelRequestMapper(
        val passengerRepository: PassengerRepository
) {

    fun map(input: TravelRequestInput): TravelRequest {
        val passenger: Passenger = passengerRepository.findById(input.passengerId)
                .orElseThrow { ResponseStatusException(HttpStatus.NOT_FOUND) }

        return TravelRequest(
                origin = input.origin,
                destination = input.destination,
                passenger = passenger
        )
    }

    fun map(travelRequest: TravelRequest) =
        TravelRequestOutput(
                creationDate = travelRequest.creationDate!!,
                destination = travelRequest.destination,
                id = travelRequest.id!!,
                origin = travelRequest.origin,
                status = travelRequest.status!!
        )

    fun buildOutputModel(travelRequest: TravelRequest, output: TravelRequestOutput): EntityModel<TravelRequestOutput> {
        val model= EntityModel<TravelRequestOutput>(output)
        travelRequest.passenger?.apply {
            val link = WebMvcLinkBuilder
                    .linkTo(PassengerAPI::class.java)
                    .slash(this.id)
                    .withRel("passenger")
                    .withTitle(this.name!!)
            model.add(link)
        }

        return model
    }

    fun buildOutputModel(requests: List<TravelRequest>) =
        requests.map { tr: TravelRequest -> buildOutputModel(tr, map(tr)) }

}